sql: restore + complete support for non-aggregated aggregate funcs

madelynnblue opened the pull request:
Aggregate functions, when called without column data, are invoked like
normal functions. That is, their `fn` property is executed. Aggregate funcs
don't include that property, so a panic was produced (for something like
`SELECT AVG(1)`). Some of these used to function properly, but that was
removed (see listed revert commit).

Restore that functionality and complete this support for the remainder of
the aggregate functions and types. Return values (including NULL) match
the behavior of postgres.

Also add decimal support to COUNT.

Partially reverts commit a55f0548af82f6899ccf6fff96dd6509d41173fa.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5654)

<!-- Reviewable:end -->

tamird commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[sql/parser/builtins.go, line 664 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5654#-KE-52WwBgQIRRWdxr-m:-KE-52WxNNIFjA5o14of:-487138053) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1ebc304356d048635190bc167b110c275e0565/sql/parser/builtins.go#L664)):*
how come, by the way?

---

*[sql/parser/builtins.go, line 1024 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5654#-KE-4YR6icspr5I-PL5w:-KE-4YR6icspr5I-PL5x:-951639409) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1ebc304356d048635190bc167b110c275e0565/sql/parser/builtins.go#L1024)):*
does the return value of this method actually matter? i struggle to see how. Perhaps is needs to be non-null, this first clause returning args[0] appears to do nothing meaningful.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5654)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod @tamird's second comment. By the way, good to hear from you Matt! 

---

Reviewed 1 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[sql/parser/builtins.go, line 664 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5654#-KE-52WwBgQIRRWdxr-m:-KE-RU3SBt1lD9HI6zNb:-1207218121) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1ebc304356d048635190bc167b110c275e0565/sql/parser/builtins.go#L664)):*
That seems to be how we defined it for multiple ints (`AVG(1, 2) = 1.5` instead of `AVG(1, 2) = 1`), so it should be consistent here. If I recall correctly, it was a deviation from Postgres.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5654#-:-KE-THBz52e4KAouvxiW:-563223504)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: restore + complete support for non-aggregated aggregate funcs

Aggregate functions, when called without column data, are invoked like
normal functions. That is, their `fn` property is executed. Aggregate funcs
don't include that property, so a panic was produced (for something like
`SELECT AVG(1)`). Some of these used to function properly, but that was
removed (see listed revert commit).

Restore that functionality and complete this support for the remainder of
the aggregate functions and types. Return values (including NULL) match
the behavior of postgres.

Also add decimal support to COUNT.

Partially reverts commit a55f0548af82f6899ccf6fff96dd6509d41173fa.

mjibson commented:
Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/parser/builtins.go, line 664 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5654#-KE-52WwBgQIRRWdxr-m:-KE-Emeceh1SuihhjjSR:41963684) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1ebc304356d048635190bc167b110c275e0565/sql/parser/builtins.go#L664)):*
I concur with Nathan. Also, this is how it was before. Obviously that doesn't make it correct, but I have no reason to believe it's wrong.

---

*[sql/parser/builtins.go, line 1024 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5654#-KE-4YR6icspr5I-PL5w:-KE2jTUx3x5lOtR8Jbvi:888583825) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1ebc304356d048635190bc167b110c275e0565/sql/parser/builtins.go#L1024)):*
This has been removed. Count is special and this fn is never invoked in any case I could test.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5654)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5654)*
<!-- Sent from Reviewable.io -->
